### ISSUE TITLE
Please switch to cloudflare for fallback DNS servers

### ISSUE THREAD

**Issue by bigon** [systemd/systemd#8899]
Hi,

In the recent years several people have shown some reservations with the fact that resolved was failing back to the Google DNS servers in some cases.

Cloudflare has started its own public DNS service a few weeks ago and for what I understand from their [announce](https://blog.cloudflare.com/announcing-1111/) is that they grantee (at least on paper) some privacy

> We began talking with browser manufacturers about what they would want from a DNS resolver. One word kept coming up: privacy. Beyond just a commitment not to use browsing data to help target ads, they wanted to make sure we would wipe all transaction logs within a week. That was an easy request. In fact, we knew we could go much further. We committed to never writing the querying IP addresses to disk and wiping all logs within 24 hours.

Shouldn't the default fallback DNS servers be switched to cloudflare ones?

The only downside (bonus so people are fixing their setup?) I'm seeing is that some ISP were/are abusing of this (previously) unallocated range for their own usage

**Comment by poettering** [comment]
So yes, @yuwata is right, the fallback DNS servers can be changed at build time and by config file, hence downstream distros as well as users have a well-defined way already to change away from the google DNS servers.

That said, I figure it might make sense to switch away from the google ones upstream too, for a variety of reasons. Question though is which provider to switch to. There's at least:

1. Google (8.8.8.8)
2. Cloudflare (1.1.1.1)
3. Quad9 (9.9.9.9)

And there are more. There's a comparison here:

https://medium.com/@nykolas.z/dns-resolvers-performance-compared-cloudflare-x-google-x-quad9-x-opendns-149e803734e5

And a wikipedia page here:

https://en.wikipedia.org/wiki/Public_recursive_name_server

I am not sure what to make of this, i.e. how to pick the best upstream one.

**Comment by bigon** [comment]
IMVHO, the default should be an "unfiltered" one.

I know that some governments are forcing providers to remove some DNS names from their DNS (ie piratebay) but using Quad9, OpenDNS or others that have an editorial line about the DNS domain they are serving is going a bit far.

**Comment by johnhtodd** [comment]
(disclaimer: I work at Quad9)  Quad9 does in fact have a malware/blocklist on 9.9.9.9/2620:fe::fe but we also operate a non-filtered resolver on 9.9.9.10 (and 2620:fe::10) which has no restrictions or filters. There are a variety of technical and non-technical reasons we think Quad9 is a good solution, but I won't enumerate them here other than to say that we're a not-for-profit that does only recursive DNS as a public service, and not as an adjunct to a commercial platform.

**Comment by shibumi** [comment]
I think there is no _best_ DNS server. They all have their up and down parts.. therefore I would just leave it as it is. Upstream can change it.

**Comment by Atavic** [comment]
Performance depends from user's location.

**Comment by davidstrauss** [comment]
I support moving from Google to Cloudflare as the default. Cloudflare's service offering has better tunneling support, retains DNSSec validation in the recursor, and introduces stronger privacy guarantees. It's also arguably faster and almost never seems to be slower (at least in data I've seen).

**Comment by shibumi** [comment]
@yuwata @poettering @davidstrauss There is another good reason why we should change to cloudflare, because of `DNS over TLS`-support. The Google DNS server don't support dns over tls, they only support dns over https. I thought it's a good idea to mention this, because of the newest changes to `systemd-resolved`.

**Comment by shibumi** [comment]
@johnhtodd Isn't quad9 a project from the British police?

**Comment by johnhtodd** [comment]
No.  https://www.quad9.net/quad9-yourdata/

**Comment by shibumi** [comment]
@johnhtodd Sure, this is one side of the medal but on the other side you have this:
```
Quad9, which has been established by IBM, the Global Security Alliance (backed by the City of London Police and Center for Internet Security) and the Packet Clearing Hous
```

And there I trust a company more than the government (sad, but that's what we got after the Snowden leaks)

**Comment by johnhtodd** [comment]
This is off topic, and your implications are entirely incorrect, and there's not much else that should be continued here to continue that thread - we've addressed this already.  DNS security and privacy are important, I agree. It is possible to imagine hypothetical and fantastic reasons that every provider that is not your own organization is doing something nefarious. If that is the case, then don't use an external provider of any sort - every single one can be accused of some unprovable misuse of data - it is never possible to prove the negative case. This thread I hope has shifted to be about what might be implemented in systemd to ensure that any options presented in the defaults (if there are even defaults) should have documented privacy policies that adhere to the most stringent (preferably GDPR in our opinion) requirements and which are open for audit, and also which enforce client encryption and other security measures (such as DNSSEC strict validation) that provide the best possible security and privacy for end users. User security/encryption, privacy standards adherence, resolver policy, and audit compliance should be the basis for inclusion in my opinion. Several public DNS services meet this bar, several do not. It seems reasonable to allow end users to make those decisions based on their own examination of the data (external links to privacy policies? a wikipedia link to comparisons?) rather than for a default to be selected without their review.

**Comment by keszybz** [comment]
> So I think it would be best to not make a choice.

We rejected this idea because it would mean that explicit configuration is always required to get a working system. Right now I can always do `meson build && ninja -C build && sudo ninja -C build install` and it might not be ideally configured, but it should mostly work.

We could make it easy to switch by providing something like built-in groups at configuration time: `@google` would give the current default, `@cloudflare` would give the IPv4 and IPv6 cloudflare servers, etc. We could even support that during run-time, i.e. allow the named groups to be used in the `DNS=` config lines.

**Comment by davidstrauss** [comment]
> There is another good reason why we should change to cloudflare, because of DNS over TLS-support.

This is what I meant by "Cloudflare's service offering has better tunneling support."

**Comment by davidstrauss** [comment]
I can think of one reason to not switch to Cloudflare DNS from the current default, and that's weirdness from certain ISPs and networks around the 1.0.0.0/8 block and 1.1.1.1 address in particular. Some networks make the incorrect assumption that those addresses are unused and either block them or handle them weirdly.

At a minimum, I would choose 1.0.0.1 for Cloudflare DNS over 1.1.1.1, as the former should exhibit marginally fewer issues with common misconfigurations.

**Comment by keszybz** [comment]
> At a minimum, I would choose 1.0.0.1 for Cloudflare DNS over 1.1.1.1

I'd expect that we chose both / all, the same as the current default is `8.8.8.8 8.8.4.4 2001:4860:4860::8888 2001:4860:4860::8844`.

**Comment by mbiebl** [comment]
@bigon fwiw, I've been going back and forth on this issue myself and I've decided to switch off the default fallback in Debian, i.e. the next Debian upload will have ```-Ddns-servers=''``` for basically the same reasons as yours.
I have to add that Ubuntu has been switching off the fallback for quite some time already:
https://salsa.debian.org/systemd-team/systemd/commit/55d514620

**Comment by shibumi** [comment]
Well, I guess this discussion is currently moving in the wrong direction. The maintainers of systemd said already that they will not remove the default DNS Servers due to stability reasons. So can we come back to the discussion about setting a default DNS server?

I would still propose cloudflare.

**Comment by mbiebl** [comment]
@shibumi what are those "stability reasons"? Personally I've been cursing on those ISPs who instead of giving me a clean error on DNS errors redirected me to a ISP specific page. The fallback mechanism is pretty close to that.

**Comment by kirbyfan64** [comment]
Since this has gone way off course, here's my third-party attempt at redirecting discussion a bit. 

To summarize benefits of CloudFlare:

- DNS over TLS support. 
- Privacy claims ("claims" because they're most likely correct, but we can't be 100% sure). 
- As far as we know CloudFlare is a mostly neutral party. 

Downsides:

- Some ISPs mess with the 1.1.1.1 and 1.0.0.1 addresses. 

Honestly given all this, a CloudFlare switch might be a good idea. Worst comes to worst, Google DNS could come after in the fallback list, that way there will be no worries of compatibility issues. 

Should @keszybz's group idea be a separate issue?

**Comment by ott** [comment]
@davidstrauss @kirbyfan64 You could use the RIPE Atlas to measure whether 1.1.1.1 would cause problems.

**Comment by davidstrauss** [comment]
@ott I looked into this by using my company's RIPE LIR membership. `8.8.8.8` performs substantially better in certain regions. Italy, for example, struggles with `1.1.1.1`:

![screenshot from 2018-10-19 12-44-48](https://user-images.githubusercontent.com/113350/47240460-39ed9d80-d39d-11e8-9f91-d34331658a36.png)

To compare, here's Italy for `8.8.8.8`:
![screenshot from 2018-10-19 12-45-59](https://user-images.githubusercontent.com/113350/47240475-45d95f80-d39d-11e8-89ae-408a36bb9fd5.png)

`1.0.0.1`, the alternative Cloudflare DNS IP, has sparser data (without commissioning a larger report) but seems better overall than `1.1.1.1`. This is why I'd advocate putting `1.0.0.1` first, then `1.1.1.1`.

![screenshot from 2018-10-19 12-47-49](https://user-images.githubusercontent.com/113350/47240521-630e2e00-d39d-11e8-8ea1-e5a9f5aa8f36.png)

In most regions, there isn't a serious difference. I focused on Italy because the results for `1.1.1.1` were notably bad compared to other regions.
